src/types/router.rs

```diff
@@ -0,0 +1,366 @@
+            nodes: &[impl AsNode],
+            constraint: f32,
+            constraint_function: fn(&dyn AsNode, &dyn AsNode) -> f32,
+            cost_function: fn(&dyn AsNode, &dyn AsNode) -> f32,
+        ) -> Router {
+            println!("[1/4] Initializing the router engine...");
+            println!("[2/4] Building edges...");
+
+            let edges = build_edges(nodes, constraint, constraint_function, cost_function);
+            let mut node_indeces = HashMap::new();
+            let mut graph = StableDiGraph::new();
+
+            println!("[3/4] Building the graph...");
+            for edge in edges {
+                let from_index = *node_indeces
+                    .entry(edge.0)
+                    .or_insert_with(|| graph.add_node(edge.0));
+                let to_index = *node_indeces
+                    .entry(edge.1)
+                    .or_insert_with(|| graph.add_node(edge.1));
+                graph.add_edge(from_index, to_index, edge.2);
+            }
+
+            println!("[4/4] Finalizing the router setup...");
+            for node in nodes {
+                if !node_indeces.contains_key(node.as_node()) {
+                    let index = graph.add_node(node.as_node());
+                    node_indeces.insert(node.as_node(), index);
+                }
+            }
+
+            println!("✨Done! Router engine is ready to use.");
+            Router {
+                graph,
+                node_indeces,
+            }
+        }
+
+        /// Get the NodeIndex struct for a given node. The NodeIndex
+        /// struct is used to reference things in the graph.
+        pub fn get_node_index(&self, node: &Node) -> Option<NodeIndex> {
+            self.node_indeces.get(node).cloned()
+        }
+
+        /// Return the number of edges in the graph.
+        pub fn get_edge_count(&self) -> usize {
+            self.graph.edge_count()
+        }
+
+        /// Find the shortest path between two nodes.
+        ///
+        /// The petgraph's dijkstra algorithm is very indentical to the
+        /// a star algorithm, so we can use the same function for both.
+        /// The only difference might be how the heuristic function is
+        /// implemented.
+        ///
+        /// # Arguments
+        /// * `from` - The node to start from.
+        /// * `to` - The node to end at.
+        /// * `algorithm` - The algorithm to use.
+        /// * `heuristic` - The heuristic function to use.
+        ///
+        /// # Returns
+        /// A tuple of the total cost and the path consisting of node
+        /// indeces.
+        ///
+        /// Note that the path excludes the ending node.
+        pub fn find_shortest_path(
+            &self,
+            from: &Node,
+            to: &Node,
+            algorithm: Algorithm,
+            heuristic_function: Option<fn(NodeIndex) -> f32>,
+        ) -> (f32, Vec<NodeIndex>) {
+            let from_index = self.get_node_index(from).unwrap();
```

Comment:
Please handle these instead of `unwrap` calls - runtime panics are not acceptable in our case

This may change the function signature and lead to bigger changers, so I would prefer to handle these as we do our PRs rather than a big cleanup later...open to discussion on this

Reply:
[f844c5a](https://github.com/Arrow-air/lib-router/pull/7/commits/f844c5a4b76fa75d82952edf83c9ac5ba96d5eaf)